Fix mount collision timeout issue

pmorie opened the pull request:
Short- or medium-term workaround for #29555.  The root issue being fixed here is that the recent attach/detach work in the kubelet uses a unique volume name as a key that tracks the work that has to be done for each volume in a pod to attach/mount/umount/detach.  However, the non-attachable volume plugins do not report unique names for themselves, which causes collisions when a single secret or configmap is mounted multiple times in a pod.

This is still a WIP -- I need to add a couple E2E tests that ensure that tests break in the future if there is a regression -- but posting for early review.

cc @kubernetes/sig-storage 

Ultimately, I would like to refine this a bit further.  A couple things I would like to change:
1.  `GetUniqueVolumeName` should be a property ONLY of attachable volumes
2.  I would like to see the kubelet apparatus for attach/mount/umount/detach handle non-attachable volumes specifically to avoid things like the `WaitForControllerAttach` call that has to be done for those volume types now

pmorie committed:
Add e2e coverage for mount collisions

pmorie commented on pkg/volume/util/volumehelper/volumehelper.go:
```diff
@@ -52,6 +52,12 @@ func GetUniqueVolumeName(pluginName, volumeName string) api.UniqueVolumeName {
+// GetUniqueVolumeNameForNonAttachableVolume returns the unique volume name
+// for a non-attachable volume.
+func GetUniqueVolumeNameForNonAttachableVolume(volumePlugin volume.VolumePlugin, podNamespace, podName, podSpecName string) api.UniqueVolumeName {
+	return api.UniqueVolumeName(fmt.Sprintf("%s/%s-%s-%s", volumePlugin.GetPluginName(), podNamespace, podName, podSpecName))
```
@jingxu97 I'm changing this to use the pod UID instead of the namespace and name to aid you in #27970

saad-ali replied:
Yep that would great

jingxu97 replied:
@pmorie, thanks for addressing this. Now I am still thinking whether this
change may affect our initial intention of avoiding operating mount/mount
for the same volume/pod. If for the same pod/volume, it is ok mount
parallel if the inner/outterspec name is different, then it should be fine.

On Wed, Jul 27, 2016 at 10:10 AM, Paul Morie <notifications@github.com>
wrote:

> In pkg/volume/util/volumehelper/volumehelper.go
> <https://github.com/kubernetes/kubernetes/pull/29673#discussion_r72480743>
> :
>
> > @@ -52,6 +52,12 @@ func GetUniqueVolumeName(pluginName, volumeName string) api.UniqueVolumeName {
> >  	return api.UniqueVolumeName(fmt.Sprintf("%s/%s", pluginName, volumeName))
> >  }
> >
> > +// GetUniqueVolumeNameForNonAttachableVolume returns the unique volume name
> > +// for a non-attachable volume.
> > +func GetUniqueVolumeNameForNonAttachableVolume(volumePlugin volume.VolumePlugin, podNamespace, podName, podSpecName string) api.UniqueVolumeName {
> > +	return api.UniqueVolumeName(fmt.Sprintf("%s/%s-%s-%s", volumePlugin.GetPluginName(), podNamespace, podName, podSpecName))
>
> @jingxu97 <https://github.com/jingxu97> I'm changing this to use the pod
> UID instead of the namespace and name to aid you in #27970
> <https://github.com/kubernetes/kubernetes/pull/27970>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29673/files/e967f0d59934d271a8eca16f6582865614008d8f#r72480743>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ASSNxfEUFfXSNMgrEIUUJwFFRtyLn6B0ks5qZ5EagaJpZM4JWQGg>
> .
>



-- 
- Jing

matchstick commented on pkg/kubelet/volumemanager/cache/desired_state_of_world.go:
```diff
@@ -183,22 +183,35 @@ func (dsw *desiredStateOfWorld) AddPodToVolume(
+			volumehelper.GetUniqueVolumeNameFromSpec(volumePlugin, volumeSpec)
+		if err != nil {
+			return "", fmt.Errorf(
+				"failed to GetUniqueVolumeNameFromSpec for volumeSpec %q using volume plugin %q err=%v",
```
@pmorie @saad-ali It might make sense to update the error to highlight that this is an attachable volume where we are failing. Just a nit.

pmorie replied:
@matchstick I think that clarification would probably only help those who would already know that if GetUniqueVolumeNameFromSpec was being called, it's necessarily an attachable volume.

matchstick commented on pkg/kubelet/volumemanager/cache/desired_state_of_world.go:
```diff
@@ -183,22 +183,35 @@ func (dsw *desiredStateOfWorld) AddPodToVolume(
+	} else {
+		// For non-attachable volumes, generate a unique name based on the pod
+		// namespace and name and the name of the volume within the pod.
+		volumeName = volumehelper.GetUniqueVolumeNameForNonAttachableVolume(volumePlugin, pod.Namespace, pod.Name, outerVolumeSpecName)
```
@pmorie @saad-ali Is it worth logging that we are now generating this name for the volume? Or is the name so clear that is not an issue? I am just curious if we may get a surprise where for some reason we have it populated in dsw.volumestomount and did not expect it. Knowing the log name we are generating may be helpful for debugging.

matchstick commented on pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go:
```diff
@@ -92,6 +92,7 @@ func (grm *nestedPendingOperations) Run(
 	grm.lock.Lock()
 	defer grm.lock.Unlock()
 
+	// search for an existing operation
```
@pmorie @saad-ali This is nice to have, but should we really bundle this in this PR? Is it unrelated? Same for the following comment.

pmorie replied:
I'll take it out.

matchstick commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -279,6 +279,11 @@ func (asw *actualStateOfWorld) MarkVolumeAsAttached(
 	return err
 }
 
+func (asw *actualStateOfWorld) MarkVolumeAsAttachedByUniqueVolumeName(
```
@saad-ali @pmorie Does it make sense to have this be public? If we could make MarkVolumeAsAttached smart enough to know when to use AddVolumeByUniqueVolumeName so users of asw should not care that is a much better solution. I think if we expect folks outside of this file to keep track of whether they need a unique name or not it is going to be setting ourselves up for someone to get it wrong.

matchstick commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -131,8 +132,14 @@ type ActualStateOfWorldMounterUpdater interface {
+	//
+	// TODO: in the future, we should be able to remove the volumeName
+	// argument to this method -- since it is used only for attachable
+	// volumes.
```
@saad-ali @pmorie Can we open the github issue and include the number in this comment?

matchstick replied:
Also add "why we do this" namely - "to ensure that the volume name used is unique."

pmorie replied:
Yes, I'll open an issue for follow-up.

pmorie committed:
Fix collisions issues / timeouts for mounts

For non-attachable volumes, do not call GetVolumeName on the plugin and instead
generate a unique name based on the identity of the pod and the name of the volume
within the pod.

pmorie commented:
@matchstick @saad-ali should be in a good state now, PTAL.

saad-ali commented:
LGTM Thanks @pmorie

fabioy commented:
@pmorie, @saad-ali  Cherrypick approved. Please create a cherrypick PR.
